neo/utils.py

```diff
@@ -0,0 +1,856 @@
+    elif any(durations == 0):
+        raise ValueError('Can not create epoch with zero duration.')
+
+    if 'name' not in kwargs:
+        kwargs['name'] = 'epoch'
+    if 'labels' not in kwargs:
+        # this needs to be changed to '%s_%i' % (kwargs['name'], i) for i in range(len(times))]
+        # when labels become unicode
+        kwargs['labels'] = [
+            ('%s_%i' % (kwargs['name'], i)).encode('ascii') for i in range(len(times))]
+
+    ep = neo.Epoch(times=times, durations=durations, **kwargs)
+
+    ep.annotate(**event1.annotations)
+
+    if attach_result:
+        segment.epochs.append(ep)
+        segment.create_relationship()
+
+    return ep
+
+
+def match_events(event1, event2):
+    """
+    Finds pairs of Event entries in event1 and event2 with the minimum delay,
+    such that the entry of event1 directly precedes the entry of event2.
+    Returns filtered two events of identical length, which contain matched
+    entries.
+
+    Parameters:
+    -----------
+    event1, event2: neo.Event
+        The two Event objects to match up.
+
+    Returns:
+    --------
+    event1, event2: neo.Event
+        Event objects with identical number of events, containing only those
+        events that could be matched against each other. A warning is issued if
+        not all events in event1 or event2 could be matched.
+    """
+    # load the full event if a proxy object has been given as an argument
+    if isinstance(event1, neo.io.proxyobjects.EventProxy):
+        event1 = event1.load()
+    if isinstance(event2, neo.io.proxyobjects.EventProxy):
+        event2 = event2.load()
+
+    id1, id2 = 0, 0
+    match_ev1, match_ev2 = [], []
+    while id1 < len(event1) and id2 < len(event2):
+        time1 = event1.times[id1]
+        time2 = event2.times[id2]
+
+        # wrong order of events
+        if time1 >= time2:
+            id2 += 1
+
+        # shorter epoch possible by later event1 entry
+        elif id1 + 1 < len(event1) and event1.times[id1 + 1] < time2:
+            # there is no event in 2 until the next event in 1
+            id1 += 1
+
+        # found a match
+        else:
+            match_ev1.append(id1)
+            match_ev2.append(id2)
+            id1 += 1
+            id2 += 1
+
+    if id1 < len(event1):
+        warnings.warn(
+            'Could not match all events to generate epochs. Missed '
+            '%s event entries in event1 list' % (len(event1) - id1))
+    if id2 < len(event2):
+        warnings.warn(
+            'Could not match all events to generate epochs. Missed '
+            '%s event entries in event2 list' % (len(event2) - id2))
+
+    event1_matched = _event_epoch_slice_by_valid_ids(
+        obj=event1, valid_ids=match_ev1)
+    event2_matched = _event_epoch_slice_by_valid_ids(
+        obj=event2, valid_ids=match_ev2)
+
+    return event1_matched, event2_matched
+
+
+def cut_block_by_epochs(block, properties=None, reset_time=False):
```

Comment:
This function seems to be doing two things: (1) getting epochs from a block (same as is done by `get_epochs()` above, no?), (2) cutting the block based on the epoch.

For maintainability, testability and consistency with `cut_segment_by_epoch()` below, I think it would be better to simplify this to only (2), so instead of 
```python
result = cut_block_by_epochs(block, some_properties)
```
you would write
```python
epochs = get_epochs(block, some_properties)
result = cut_block_by_epochs(block, epochs)

Reply:
The idea behind this is that a Segment is only cut by Epochs which belong to that Segment. For each Segment, we have to go through all of its Epochs and filter and slice them according to the requested properties. Only these selected Epochs are used to cut that same Segment. If the Epoch extraction is a separate step we have to somehow keep track of which Segment each Epoch comes from. Since the extracted Epochs are duplicates or slices of existing ones they are not attached to the neo structure and this metadata is lost.